src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs

```diff
@@ -79,17 +80,42 @@ public ArrayBufferWriter(int initialCapacity)
 
         /// <summary>
         /// Clears the data written to the underlying buffer.
+        /// When <c>T</c> is not reference and not contains reference,
+        /// using <seealso cref="ResetWrittenCount"/> is more performance effective.
         /// </summary>
         /// <remarks>
-        /// You must clear the <see cref="ArrayBufferWriter{T}"/> before trying to re-use it.
+        /// You must reset or clear the <see cref="ArrayBufferWriter{T}"/> before trying
+        /// to re-use it.
         /// </remarks>
+        /// <seealso cref="ResetWrittenCount"/>
+        /// <seealso cref="RuntimeHelpers.IsReferenceOrContainsReferences{T}"/>
         public void Clear()
         {
             Debug.Assert(_buffer.Length >= _index);
             _buffer.AsSpan(0, _index).Clear();
             _index = 0;
         }
 
+        /// <summary>
+        /// Resets the data written to the underlying buffer without zeroing memory
+        /// when <c>T</c> is not reference and not contains reference, otherwise clears them.
+        /// </summary>
+        /// <remarks>
+        /// You must reset or clear the <see cref="ArrayBufferWriter{T}"/> before trying
+        /// to re-use it.
+        /// </remarks>
+        /// <seealso cref="Clear"/>
+        /// <seealso cref="RuntimeHelpers.IsReferenceOrContainsReferences{T}"/>
+        public void ResetWrittenCount()
+        {
+            Debug.Assert(_buffer.Length >= _index);
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
```

Comment:
@terrajobst, @bartonjs, @davidfowl, if the only difference between Clear and ResetWrittenCount is whether the memory is zeroed, would it makes sense to _not_ special case references and just always not zero? That enables a caller to skip zeroing even for references if that's the behavior they want, or to themselves choose whether to Clear or ResetWrittenCount based on whether the T has references. Basically, should this entire method just be `_index = 0`?

Reply:
If you do not clear the references, then such objects will not be released immediately, but only together with the buffer. If the buffer is cached somewhere for a long time, a memory leak may occur. I do not know how this situation is possible and whether it makes sense then to add another method for clearing the entire buffer.

Reply:
> If you do not clear the references, then such objects will not be released immediately

That is why Clear clears. The only benefit to the new API is that it doesn't clear. If for things with references we still always clear, the new API has no benefit. Hence my question.

> I do not know how this situation is possible and whether it makes sense then to add another method for clearing the entire buffer.

I don't think that's desirable. You can always just drop the ArrayBufferWriter itself. And my point was that by having this not zero, we give the consumer the flexibility.

The only argument I see for zeroing here as well is basically "The 99.9% use case for this type is with bytes and chars, and that's where perf matters; for the remaining 0.1% that might include references, we're sure the cost of zeroing doesn't matter and we want to make the API safer."

Reply:
I suggest the next solution:
1. This method always resets written count and does not clear memory.
2. To clear all buffer memory next code can be used in user code:

```cs
    buffer.Advance(buffer.FreeCapacity);
    buffer.Clear();
```

Reply:
@terrajobst, @bartonjs, @davidfowl, are you ok with where we landed here, that the entirety of the method is:
```C#
public void ResetWrittenCount() => _index = 0;
```
even if T is or contains references? I think it makes sense, but I want to make sure others agree.

Reply:
My only issue with it is that you can't recover from it by calling Clear.  You have to either advance back to the full position and then call Clear, or drop the whole object.

It probably won't end up mattering, since I didn't see anything on grep.app using ArrayBufferWriter for anything other than byte or char.  So, in that sense, I'm OK with it, I suppose :)